### DISCUSSION TITLE
Can you please explain why in public clients, redirects are very important?

### DISCUSSION THREAD

**Discussion by Nefcanto** [keycloak/keycloak#35973] [Q&A]
The [doc](https://www.keycloak.org/securing-apps/javascript-adapter#_keycloak_server_configuration) says:

> One important thing to consider about using client-side applications is that the client has to be a public client as there is no secure way to store client credentials in a client-side application. This consideration makes it very important to make sure the redirect URIs you have configured for the client are correct and as specific as possible.

Why is it so important in this case?

**Comment by bonnm** [comment]
An attacker can construct an authorization URL using a client ID that belongs to a known good application, but set the redirect URL to a URL under the control of the attacker. If the authorization server does not validate redirect URLs, and the attacker uses the “token” response type (implicit flow), the user will be returned to the attacker’s application with the access token in the URL. If the client is a public client, and the attacker intercepts the authorization code (code flow), then the attacker can also exchange the code for an access token.

Especially in public cloud or web hosting environments you should be careful when using the redirect uri wildcard option of Keycloak.

**Reply by embesozzi** [reply]
The point is that public clients don't have a client secret because there is no backend in SPAs, and therefore no way to store a secret securely. Consequently, there is no authentication of the client when negotiating the token.
This is why, in these cases, you end up using Authorization Code + PKCE to help prevent CSRF and authorization code injection attacks.

**Reply by darius-m** [reply]
A small correction: if I understand its purpose correctly, PKCE is meant to protect against code intercepts from applications on the same device (e.g., a default URL handler registered on the device that redirects the request to a rogue application), as only the application that generated the request would know the nonce sent in the original authentication request.

Other than that, for public clients it's what everyone else has said - the SSO server has no way of validating the backend that requests user tokens (the secret otherwise works as a form of username+password authentication). If an adversary can make the user click on a link that contains a post-authentication redirect link to a domain controlled by the attacker, they could use the code to fetch the user token and authenticate as the user to other services. Of course, this depends on the information in the token, especially the audience field, and whether it is validated by the other services. Having strict origin and redirect URI whitelists can help mitigate this issue, as the SSO can refuse to redirect the user to untrusted domains, or if the request does not have the correct origin.
